File: kit/src/statis_kit/cost_meter.py
"""Token counting and cost estimation — tiktoken optional."""
from __future__ import annotations

from typing import Callable, Optional

from ._pricing import get_model_pricing, load_pricing
from ._types import CostEstimate, MeterConfig, Message, TurnCost


def _get_counter(model: str) -> Callable[[str], int]:
    """Return a token-counting function. Uses tiktoken when available."""
    try:
        import tiktoken  # type: ignore[import]

        # Map model names to tiktoken encodings
        try:
            enc = tiktoken.encoding_for_model(model)
        except KeyError:
            enc = tiktoken.get_encoding("cl100k_base")

        def _count(text: str) -> int:
            return len(enc.encode(text))

        return _count
    except ImportError:
        pass

    # Fallback: chars / 4 (rough approximation for English text)
    def _fallback(text: str) -> int:
        return max(1, len(text) // 4)

    return _fallback


class CostMeter:
    def __init__(self, config: Optional[MeterConfig] = None) -> None:
        cfg = config or MeterConfig()
        self._model = cfg.model
        self._pricing = load_pricing(cfg.pricing_path)
        self._on_turn = cfg.on_turn
        self._counter = _get_counter(cfg.model)

    def count_tokens(self, text: str) -> int:
        """Count tokens in a single string."""
        return self._counter(text)

    def count_messages(
        self, messages: list[Message],
    ) -> tuple[int, list[TurnCost]]:
        """Count tokens per turn and session total.

        Returns (total_tokens, per_turn_costs).
        """
        input_per_1k, _ = get_model_pricing(self._model, self._pricing)
        per_turn: list[TurnCost] = []
        total = 0

        for i, msg in enumerate(messages):
            tokens = self._counter(msg.content)
            cost = (tokens / 1000.0) * input_per_1k
            tc = TurnCost(
                turn_index=i,
                role=msg.role,
                tokens=tokens,
                cost_usd=cost,
            )
            per_turn.append(tc)
            total += tokens

            if self._on_turn is not None:
                self._on_turn(tc)

        return total, per_turn

    def estimate_cost(
        self,
        input_tokens: int,
        output_tokens: int = 0,
    ) -> CostEstimate:
        """Estimate cost based on model pricing table."""
        input_per_1k, output_per_1k = get_model_pricing(
            self._model, self._pricing,
        )
        input_cost = (input_tokens / 1000.0) * input_per_1k
        output_cost = (output_tokens / 1000.0) * output_per_1k
        return CostEstimate(
            input_cost_usd=round(input_cost, 8),
            output_cost_usd=round(output_cost, 8),
            total_cost_usd=round(input_cost + output_cost, 8),
            model=self._model,
        )

File: kit-ts/src/cost-meter.ts
import type { CostEstimate, MeterConfig, Message, ModelPricing, PricingTable, TurnCost } from "./types";
import DEFAULT_PRICING from "../data/pricing.json";

function loadPricing(path?: string): PricingTable {
  if (path === undefined) {
    return DEFAULT_PRICING as PricingTable;
  }
  // Custom path — Node-only. Dynamic require keeps browser bundles clean.
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { readFileSync } = require("node:fs");
  return JSON.parse(readFileSync(path, "utf-8")) as PricingTable;
}

function getModelPricing(model: string, pricing?: PricingTable): ModelPricing {
  const table = pricing ?? loadPricing();
  const models = table.models;

  if (model in models) return models[model];

  // Bidirectional prefix match — matches both dated→short and short→dated forms
  for (const key of Object.keys(models)) {
    if (model.startsWith(key) || key.startsWith(model)) return models[key];
  }

  // Default to gpt-4o
  return models["gpt-4o"] ?? { input_per_1k: 0.0025, output_per_1k: 0.01 };
}

function countTokensFallback(text: string): number {
  return Math.max(1, Math.floor(text.length / 4));
}

export class CostMeter {
  private readonly model: string;
  private readonly pricing: PricingTable;
  private readonly onTurn?: (cost: TurnCost) => void;
  private readonly counter: (text: string) => number;

  constructor(config?: MeterConfig) {
    const cfg = config ?? {};
    this.model = cfg.model ?? "gpt-4o";
    this.pricing = loadPricing(cfg.pricing_path);
    this.onTurn = cfg.on_turn;
    this.counter = countTokensFallback;
  }

  countTokens(text: string): number {
    return this.counter(text);
  }

  countMessages(messages: Message[]): { total: number; perTurn: TurnCost[] } {
    const { input_per_1k } = getModelPricing(this.model, this.pricing);
    const perTurn: TurnCost[] = [];
    let total = 0;

    for (let i = 0; i < messages.length; i++) {
      const msg = messages[i];
      const tokens = this.counter(msg.content);
      const cost_usd = (tokens / 1000) * input_per_1k;
      const tc: TurnCost = { turn_index: i, role: msg.role, tokens, cost_usd };
      perTurn.push(tc);
      total += tokens;
      this.onTurn?.(tc);
    }

    return { total, perTurn };
  }

  estimateCost(inputTokens: number, outputTokens: number = 0): CostEstimate {
    const pricing = getModelPricing(this.model, this.pricing);
    const inputCost = (inputTokens / 1000) * pricing.input_per_1k;
    const outputCost = (outputTokens / 1000) * pricing.output_per_1k;
    return {
      input_cost_usd: parseFloat(inputCost.toFixed(8)),
      output_cost_usd: parseFloat(outputCost.toFixed(8)),
      total_cost_usd: parseFloat((inputCost + outputCost).toFixed(8)),
      model: this.model,
    };
  }
}
